### PR TITLE
Add illumos triple

### LINKED ISSUE #55553: Create a separate target for illumos
Illumos seems to be becoming less and less similar to its ancestor – Solaris. There are multiple incompatibilities at the OS interface level – https://github.com/rust-random/rand/issues/637 and https://github.com/rust-lang/rust/issues/52577 are the things that come to mind, but I’m sure there are more and the differences will keep coming up more and more frequently as the two diverge over time.

### PR DESCRIPTION
**Pull request by pfmooney**
This fixes rust-lang/rust#55553 and adds support for `illumos` as a `target_os` on `x86_64`.  In addition to the compile spec and libstd additions, several library dependencies have been bumped in order to permit working builds of cargo and rustup for the new target.

Work originally started by @jasonbking, with subsequent additions by @pfmooney and @jclulow.

### CODE REVIEW

**Comment by pfmooney** [comment]
The cargo update in #71138 will supersede the need for one here, assuming it lands first.

**Comment by pfmooney** [comment]
Rebased on top of #71138 and updated to include `set_nonblocking()` modification for `Socket` on illumos and Solaris.

**Comment by pfmooney** [comment]
cc: @nagisa

File: src/libstd/sys/unix/net.rs [hunk 1]
```diff
@@ -322,11 +322,19 @@ impl Socket {
...
     pub fn set_nonblocking(&self, nonblocking: bool) -> io::Result<()> {
         let mut nonblocking = nonblocking as libc::c_int;
         cvt(unsafe { libc::ioctl(*self.as_inner(), libc::FIONBIO, &mut nonblocking) }).map(drop)
     }
 
+    #[cfg(any(target_os = "solaris", target_os = "illumos"))]
+    pub fn set_nonblocking(&self, nonblocking: bool) -> io::Result<()> {
+        // FIONBIO is inadequate for sockets on illumos/Solaris, so use the
+        // fcntl(F_[GS]ETFL)-based method provided by FileDesc instead.
+        self.0.set_nonblocking(nonblocking)
```
Comment (RalfJung) [inline src/libstd/sys/unix/net.rs:335]: This generally [seems to be the portable and recommended method](https://stackoverflow.com/questions/1150635/unix-nonblocking-i-o-o-nonblock-vs-fionbio) on POSIX. Makes me wonder why we do something else on Linux and other Unices.

**Pull request merged**